// @devvit/protos/json/devvit/dev_portal/app_publish_request/app_publish_request.js
/**
 * #app_publish_request.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var AppPublishRequestStatus;
(function (AppPublishRequestStatus) {
    AppPublishRequestStatus[AppPublishRequestStatus["APPROVED"] = 0] = "APPROVED";
    AppPublishRequestStatus[AppPublishRequestStatus["DENIED"] = 1] = "DENIED";
    AppPublishRequestStatus[AppPublishRequestStatus["OBSOLETE"] = 2] = "OBSOLETE";
    AppPublishRequestStatus[AppPublishRequestStatus["PENDING"] = 3] = "PENDING";
    AppPublishRequestStatus[AppPublishRequestStatus["WITHDRAWN"] = 4] = "WITHDRAWN";
    AppPublishRequestStatus[AppPublishRequestStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AppPublishRequestStatus || (AppPublishRequestStatus = {}));
export var AppPublishRequestVisibility;
(function (AppPublishRequestVisibility) {
    AppPublishRequestVisibility[AppPublishRequestVisibility["PUBLIC"] = 0] = "PUBLIC";
    AppPublishRequestVisibility[AppPublishRequestVisibility["UNLISTED"] = 1] = "UNLISTED";
    AppPublishRequestVisibility[AppPublishRequestVisibility["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AppPublishRequestVisibility || (AppPublishRequestVisibility = {}));
/**
 * AppPublishRequestPendingState describes the queue context for a pending
 * app publish request. This metadata is secondary to the primary request
 * status and is used to clarify why a request is still in review.
 */
export var AppPublishRequestPendingState;
(function (AppPublishRequestPendingState) {
    AppPublishRequestPendingState[AppPublishRequestPendingState["AWAITING_REVIEW"] = 0] = "AWAITING_REVIEW";
    AppPublishRequestPendingState[AppPublishRequestPendingState["AWAITING_PANEL_REVIEW"] = 1] = "AWAITING_PANEL_REVIEW";
    AppPublishRequestPendingState[AppPublishRequestPendingState["IN_REVIEW"] = 2] = "IN_REVIEW";
    AppPublishRequestPendingState[AppPublishRequestPendingState["ON_HOLD"] = 3] = "ON_HOLD";
    AppPublishRequestPendingState[AppPublishRequestPendingState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AppPublishRequestPendingState || (AppPublishRequestPendingState = {}));
export var AutomatedAppReviewDecision;
(function (AutomatedAppReviewDecision) {
    AutomatedAppReviewDecision[AutomatedAppReviewDecision["AUTOMATED_APP_REVIEW_DECISION_UNSPECIFIED"] = 0] = "AUTOMATED_APP_REVIEW_DECISION_UNSPECIFIED";
    AutomatedAppReviewDecision[AutomatedAppReviewDecision["AUTOMATED_APP_REVIEW_DECISION_APPROVED"] = 1] = "AUTOMATED_APP_REVIEW_DECISION_APPROVED";
    AutomatedAppReviewDecision[AutomatedAppReviewDecision["AUTOMATED_APP_REVIEW_DECISION_REJECTED"] = 2] = "AUTOMATED_APP_REVIEW_DECISION_REJECTED";
    AutomatedAppReviewDecision[AutomatedAppReviewDecision["AUTOMATED_APP_REVIEW_DECISION_NEEDS_REVIEW"] = 3] = "AUTOMATED_APP_REVIEW_DECISION_NEEDS_REVIEW";
    AutomatedAppReviewDecision[AutomatedAppReviewDecision["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AutomatedAppReviewDecision || (AutomatedAppReviewDecision = {}));
export var AutomatedAppReviewStatus;
(function (AutomatedAppReviewStatus) {
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["AUTOMATED_APP_REVIEW_STATUS_UNSPECIFIED"] = 0] = "AUTOMATED_APP_REVIEW_STATUS_UNSPECIFIED";
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["AUTOMATED_APP_REVIEW_STATUS_PENDING"] = 1] = "AUTOMATED_APP_REVIEW_STATUS_PENDING";
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["AUTOMATED_APP_REVIEW_STATUS_COMPLETED"] = 2] = "AUTOMATED_APP_REVIEW_STATUS_COMPLETED";
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["AUTOMATED_APP_REVIEW_STATUS_ERROR"] = 3] = "AUTOMATED_APP_REVIEW_STATUS_ERROR";
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["AUTOMATED_APP_REVIEW_STATUS_RATE_LIMITED"] = 4] = "AUTOMATED_APP_REVIEW_STATUS_RATE_LIMITED";
    AutomatedAppReviewStatus[AutomatedAppReviewStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AutomatedAppReviewStatus || (AutomatedAppReviewStatus = {}));
