// @devvit/protos/json/devvit/dev_portal/installation/installation.js
/**
 * #installation.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var UpgradeStrategy;
(function (UpgradeStrategy) {
    UpgradeStrategy[UpgradeStrategy["MANUAL"] = 0] = "MANUAL";
    UpgradeStrategy[UpgradeStrategy["AUTOMATIC"] = 1] = "AUTOMATIC";
    UpgradeStrategy[UpgradeStrategy["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(UpgradeStrategy || (UpgradeStrategy = {}));
/**
 * Permissions that a moderator can grant to the app developer for an installation
 * This enum maps 1-to-1 to its Prisma equivalent in prisma/schema.prisma,
 * so they must be kept in sync
 */
export var DeveloperPermissions;
(function (DeveloperPermissions) {
    /** NONE - No permissions granted */
    DeveloperPermissions[DeveloperPermissions["NONE"] = 0] = "NONE";
    /**
     * READ_INSTALL_LOGS - Developer can stream the installation's logs and see the installation page
     * in portal, but cannot edit the installation's settings
     */
    DeveloperPermissions[DeveloperPermissions["READ_INSTALL_LOGS"] = 1] = "READ_INSTALL_LOGS";
    DeveloperPermissions[DeveloperPermissions["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DeveloperPermissions || (DeveloperPermissions = {}));
/**
 * State of an app installation
 * This enum maps 1-to-1 to its Prisma equivalent in prisma/schema.prisma,
 * so they must be kept in sync
 */
export var AppInstallState;
(function (AppInstallState) {
    /** ACTIVE - Installation is active and running */
    AppInstallState[AppInstallState["ACTIVE"] = 0] = "ACTIVE";
    /** DISABLED - Installation is disabled (not deleted, just turned off) */
    AppInstallState[AppInstallState["DISABLED"] = 1] = "DISABLED";
    /** APP_UNINSTALLED - Installation has been uninstalled (soft deleted) */
    AppInstallState[AppInstallState["APP_UNINSTALLED"] = 2] = "APP_UNINSTALLED";
    AppInstallState[AppInstallState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AppInstallState || (AppInstallState = {}));
export var InstallationHistoryEventName;
(function (InstallationHistoryEventName) {
    InstallationHistoryEventName[InstallationHistoryEventName["INSTALLED"] = 0] = "INSTALLED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNINSTALLED"] = 1] = "UNINSTALLED";
    InstallationHistoryEventName[InstallationHistoryEventName["UPGRADED"] = 2] = "UPGRADED";
    InstallationHistoryEventName[InstallationHistoryEventName["UPDATED"] = 3] = "UPDATED";
    InstallationHistoryEventName[InstallationHistoryEventName["BANNED"] = 4] = "BANNED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNBANNED"] = 5] = "UNBANNED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(InstallationHistoryEventName || (InstallationHistoryEventName = {}));
