// @devvit/public-api/apis/reddit/models/User.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _User_id, _User_username, _User_createdAt, _User_linkKarma, _User_commentKarma, _User_nsfw, _User_isAdmin, _User_isModerator, _User_isGold, _User_modPermissionsBySubreddit, _User_showNsfw, _User_url, _User_permalink, _User_hasVerifiedEmail, _User_displayName, _User_about, _User_metadata;
import { Header } from '@devvit/shared-types/Header.js';
import { assertNonNull } from '@devvit/shared-types/NonNull.js';
import { Devvit } from '../../../devvit/Devvit.js';
import { asT2ID, asT6ID, isT2ID, isT6ID } from '../../../types/tid.js';
import { GraphQL } from '../graphql/GraphQL.js';
import { makeGettersEnumerable } from '../helpers/makeGettersEnumerable.js';
import { formatModeratorPermissions, validModPermissions } from '../helpers/permissions.js';
import { Comment } from './Comment.js';
import { convertUserFlairProtoToAPI, Flair } from './Flair.js';
import { Listing } from './Listing.js';
import { Post } from './Post.js';
export var SocialLinkType;
(function (SocialLinkType) {
    SocialLinkType["Custom"] = "CUSTOM";
    SocialLinkType["Reddit"] = "REDDIT";
    SocialLinkType["Instagram"] = "INSTAGRAM";
    SocialLinkType["Twitter"] = "TWITTER";
    SocialLinkType["Tiktok"] = "TIKTOK";
    SocialLinkType["Twitch"] = "TWITCH";
    SocialLinkType["Facebook"] = "FACEBOOK";
    SocialLinkType["Youtube"] = "YOUTUBE";
    SocialLinkType["Tumblr"] = "TUMBLR";
    SocialLinkType["Spotify"] = "SPOTIFY";
    SocialLinkType["Soundcloud"] = "SOUNDCLOUD";
    SocialLinkType["Beacons"] = "BEACONS";
    SocialLinkType["Linktree"] = "LINKTREE";
    SocialLinkType["Discord"] = "DISCORD";
    SocialLinkType["Venmo"] = "VENMO";
    SocialLinkType["CashApp"] = "CASH_APP";
    SocialLinkType["Patreon"] = "PATREON";
    SocialLinkType["Kofi"] = "KOFI";
    SocialLinkType["Paypal"] = "PAYPAL";
    SocialLinkType["Cameo"] = "CAMEO";
    SocialLinkType["Onlyfans"] = "ONLYFANS";
    SocialLinkType["Substack"] = "SUBSTACK";
    SocialLinkType["Kickstarter"] = "KICKSTARTER";
    SocialLinkType["Indiegogo"] = "INDIEGOGO";
    SocialLinkType["BuyMeACoffee"] = "BUY_ME_A_COFFEE";
    SocialLinkType["Shopify"] = "SHOPIFY";
})(SocialLinkType || (SocialLinkType = {}));
/**
 * A class representing a user.
 */
export class User {
    /**
     * @internal
     */
    constructor(data, metadata) {
        _User_id.set(this, void 0);
        _User_username.set(this, void 0);
        _User_createdAt.set(this, void 0);
        _User_linkKarma.set(this, void 0);
        _User_commentKarma.set(this, void 0);
        _User_nsfw.set(this, void 0);
        _User_isAdmin.set(this, void 0);
        _User_isModerator.set(this, void 0);
        _User_isGold.set(this, void 0);
        _User_modPermissionsBySubreddit.set(this, new Map());
        _User_showNsfw.set(this, void 0);
        // R2 bug: user.url is a permalink path
        _User_url.set(this, void 0);
        // R2 bug: user object does not contain a permalink field
        _User_permalink.set(this, void 0);
        _User_hasVerifiedEmail.set(this, void 0);
        _User_displayName.set(this, void 0);
        _User_about.set(this, void 0);
        _User_metadata.set(this, void 0);
        makeGettersEnumerable(this);
        assertNonNull(data.id, 'User ID is missing or undefined');
        assertNonNull(data.name, 'Username is missing or undefined');
        assertNonNull(data.createdUtc, 'User is missing created date');
        // UserDataByAccountIds returns the ID without the t2_ prefix
        __classPrivateFieldSet(this, _User_id, asT2ID(isT2ID(data.id) ? data.id : `t2_${data.id}`), "f");
        __classPrivateFieldSet(this, _User_username, data.name, "f");
        __classPrivateFieldSet(this, _User_nsfw, data.subreddit?.over18 ?? false, "f");
        __classPrivateFieldSet(this, _User_isAdmin, data.isEmployee ?? false, "f");
        __classPrivateFieldSet(this, _User_isModerator, data.isMod ?? false, "f");
        __classPrivateFieldSet(this, _User_isGold, data.isGold ?? false, "f");
        __classPrivateFieldSet(this, _User_showNsfw, data.over18 ?? false, "f");
        const createdAt = new Date(0);
        createdAt.setUTCSeconds(data.createdUtc);
        __classPrivateFieldSet(this, _User_createdAt, createdAt, "f");
        __classPrivateFieldSet(this, _User_linkKarma, data.linkKarma ?? 0, "f");
        __classPrivateFieldSet(this, _User_commentKarma, data.commentKarma ?? 0, "f");
        if (data.modPermissions) {
            for (const [subredditName, permissions] of Object.entries(data.modPermissions)) {
                __classPrivateFieldGet(this, _User_modPermissionsBySubreddit, "f").set(subredditName, validModPermissions(permissions));
            }
        }
        __classPrivateFieldSet(this, _User_url, new URL(data.subreddit?.url ?? '', 'https://www.reddit.com').toString(), "f");
        __classPrivateFieldSet(this, _User_permalink, data.subreddit?.url ?? '', "f");
        __classPrivateFieldSet(this, _User_hasVerifiedEmail, data.hasVerifiedEmail ?? false, "f");
        __classPrivateFieldSet(this, _User_displayName, data.subreddit?.title ?? __classPrivateFieldGet(this, _User_username, "f"), "f");
        __classPrivateFieldSet(this, _User_about, data.subreddit?.publicDescription ?? '', "f");
        __classPrivateFieldSet(this, _User_metadata, metadata, "f");
    }
    /**
     * The ID (starting with t2_) of the user to retrieve.
     * @example 't2_1w72'
     */
    get id() {
        return __classPrivateFieldGet(this, _User_id, "f");
    }
    /**
     * The username of the user omitting the u/.
     * @example 'spez'
     */
    get username() {
        return __classPrivateFieldGet(this, _User_username, "f");
    }
    /**
     * The date the user was created.
     */
    get createdAt() {
        return __classPrivateFieldGet(this, _User_createdAt, "f");
    }
    /**
     * The amount of link karma the user has.
     */
    get linkKarma() {
        return __classPrivateFieldGet(this, _User_linkKarma, "f");
    }
    /**
     * The amount of comment karma the user has.
     */
    get commentKarma() {
        return __classPrivateFieldGet(this, _User_commentKarma, "f");
    }
    /**
     * Whether the user's profile is marked as NSFW (Not Safe For Work).
     */
    get nsfw() {
        return __classPrivateFieldGet(this, _User_nsfw, "f");
    }
    /**
     * Whether the user is admin.
     */
    get isAdmin() {
        return __classPrivateFieldGet(this, _User_isAdmin, "f");
    }
    /** Whether the user is a moderator of any subreddit. */
    get isModerator() {
        return __classPrivateFieldGet(this, _User_isModerator, "f");
    }
    /** Whether the user has Reddit Premium. */
    get hasRedditPremium() {
        return __classPrivateFieldGet(this, _User_isGold, "f");
    }
    /**
     * The permissions the user has on the subreddit.
     */
    get modPermissions() {
        return __classPrivateFieldGet(this, _User_modPermissionsBySubreddit, "f");
    }
    /** Whether the user is over 18 and wishes to see NSFW content.*/
    get showNsfw() {
        return __classPrivateFieldGet(this, _User_showNsfw, "f");
    }
    /**
     * Returns the HTTP URL for the user
     */
    get url() {
        return __classPrivateFieldGet(this, _User_url, "f");
    }
    /**
     * Returns a permalink path relative to https://www.reddit.com
     */
    get permalink() {
        return __classPrivateFieldGet(this, _User_permalink, "f");
    }
    /**
     * Indicates whether or not the user has verified their email address.
     */
    get hasVerifiedEmail() {
        return __classPrivateFieldGet(this, _User_hasVerifiedEmail, "f");
    }
    /**
     * The display name of the user. May be different from their username.
     */
    get displayName() {
        return __classPrivateFieldGet(this, _User_displayName, "f");
    }
    /**
     * The user's public description about themselves. May be empty.
     */
    get about() {
        return __classPrivateFieldGet(this, _User_about, "f");
    }
    toJSON() {
        return {
            id: this.id,
            username: this.username,
            createdAt: this.createdAt,
            linkKarma: this.linkKarma,
            commentKarma: this.commentKarma,
            nsfw: this.nsfw,
            modPermissionsBySubreddit: Object.fromEntries(this.modPermissions),
        };
    }
    /**
     * Get the mod permissions the user has on the subreddit if they are a moderator.
     *
     * @param subredditName - name of the subreddit
     * @returns the moderator permissions the user has on the subreddit
     */
    async getModPermissionsForSubreddit(subredditName) {
        if (__classPrivateFieldGet(this, _User_modPermissionsBySubreddit, "f").has(subredditName)) {
            return __classPrivateFieldGet(this, _User_modPermissionsBySubreddit, "f").get(subredditName);
        }
        const mods = await User.getSubredditUsersByType({
            subredditName,
            type: 'moderators',
            username: this.username,
        }, __classPrivateFieldGet(this, _User_metadata, "f")).all();
        if (mods.length === 0) {
            return [];
        }
        const permissions = mods[0].modPermissions.get(subredditName) ?? [];
        if (permissions.length > 0) {
            __classPrivateFieldGet(this, _User_modPermissionsBySubreddit, "f").set(subredditName, permissions);
        }
        return permissions;
    }
    /**
     * Get the user's comments.
     *
     * @param options - Options for the request
     * @param options.sort - The sort order of the comments. e.g. 'new'
     * @param options.timeframe - The timeframe of the comments. e.g. 'all'
     * @param options.limit - The maximum number of comments to return. e.g. 1000
     * @param options.pageSize - The number of comments to return per request. e.g. 100
     * @returns A Listing of Comment objects.
     */
    getComments(options) {
        return Comment.getCommentsByUser({
            username: this.username,
            ...options,
        }, __classPrivateFieldGet(this, _User_metadata, "f"));
    }
    /**
     * Get the user's posts.
     *
     * @param options - Options for the request
     * @param options.sort - The sort order of the posts. e.g. 'new'
     * @param options.timeframe - The timeframe of the posts. e.g. 'all'
     * @param options.limit - The maximum number of posts to return. e.g. 1000
     * @param options.pageSize - The number of posts to return per request. e.g. 100
     * @returns A Listing of Post objects.
     */
    getPosts(options) {
        return Post.getPostsByUser({
            username: this.username,
            ...options,
        }, __classPrivateFieldGet(this, _User_metadata, "f"));
    }
    /**
     * Retrieve the user's flair for the subreddit.
     *
     * @param subreddit - The name of the subreddit associated with the user's flair.
     *
     * @example
     * ```ts
     * const username = "badapple"
     * const subredditName = "mysubreddit"
     * const user = await reddit.getUserByUsername(username);
     * const userFlair = await user.getUserFlairBySubreddit(subredditName);
     * ```
     */
    async getUserFlairBySubreddit(subreddit) {
        const userFlairs = await Flair.getUserFlairBySubreddit({
            subreddit,
            name: __classPrivateFieldGet(this, _User_username, "f"),
        }, __classPrivateFieldGet(this, _User_metadata, "f"));
        return userFlairs.users[0] ? convertUserFlairProtoToAPI(userFlairs.users[0]) : undefined;
    }
    getSnoovatarUrl() {
        return User.getSnoovatarUrl(this.username, __classPrivateFieldGet(this, _User_metadata, "f"));
    }
    /**
     * Gets social links of the user
     *
     * @returns A Promise that resolves an Array of UserSocialLink objects
     * @example
     * ```ts
     * const socialLinks = await user.getSocialLinks();
     * ```
     */
    async getSocialLinks() {
        const operationName = 'GetUserSocialLinks';
        const persistedQueryHash = '2aca18ef5f4fc75fb91cdaace3e9aeeae2cb3843b5c26ad511e6f01b8521593a';
        // Legacy GQL query. Do not copy this pattern.
        // eslint-disable-next-line no-restricted-properties
        const response = await GraphQL.query(operationName, persistedQueryHash, { name: this.username }, __classPrivateFieldGet(this, _User_metadata, "f"));
        if (!response.data?.user?.profile?.socialLinks) {
            return [];
        }
        return response.data.user.profile.socialLinks.map((link) => ({
            ...link,
            handle: link.handle ?? undefined,
        }));
    }
    /**
     * Returns the karma for this User in the current subreddit.
     * The user making the request must be a moderator of the subreddit to read another user's karma in the subreddit.
     * An exception is if the specified user is the same as the user making the request.
     *
     * @returns The GetUserKarmaForSubredditResponse, containing the user's karma for comments and posts in the subreddit.
     */
    async getUserKarmaFromCurrentSubreddit() {
        return await User.getUserKarmaFromCurrentSubreddit(this.username, __classPrivateFieldGet(this, _User_metadata, "f"));
    }
    /**
     * Get the trophies displayed on this user's profile.
     *
     * @returns A Promise that resolves to an array of Trophy objects.
     */
    async getTrophies() {
        return User.getTrophies(this.username, __classPrivateFieldGet(this, _User_metadata, "f"));
    }
    /** @internal */
    static async getById(id, metadata) {
        const username = await getUsernameById(id, metadata);
        return username == null ? undefined : User.getByUsername(username, metadata);
    }
    /** @internal */
    static async getByUsername(username, metadata) {
        const client = Devvit.redditAPIPlugins.Users;
        try {
            const response = await client.UserAbout({ username }, metadata);
            // suspended accounts 404.
            if (response.data?.id)
                return new User(response.data, metadata);
        }
        catch (error) {
            if (error instanceof Error && error.message.includes('404 Not Found')) {
                return undefined;
            }
            throw error;
        }
    }
    /** @internal */
    static async getFromMetadata(key, metadata) {
        assertNonNull(metadata);
        const userId = metadata?.[key]?.values[0];
        return userId ? User.getById(asT2ID(userId), metadata) : Promise.resolve(undefined);
    }
    /** @internal */
    static getSubredditUsersByType(options, metadata) {
        const client = Devvit.redditAPIPlugins.Subreddits;
        return new Listing({
            hasMore: true,
            pageSize: options.pageSize,
            limit: options.limit,
            after: options.after,
            before: options.before,
            fetch: async (fetchOptions) => {
                const response = await client.AboutWhere({
                    where: options.type,
                    user: options.username,
                    subreddit: options.subredditName,
                    show: 'all',
                    ...fetchOptions,
                }, metadata);
                return listingProtosToUsers(response, options.subredditName, metadata);
            },
        });
    }
    /** @internal */
    static async createRelationship(options, metadata) {
        const client = Devvit.redditAPIPlugins.Users;
        const { type, subredditName, username, permissions, ...optionalFields } = options;
        const response = await client.Friend({
            type,
            subreddit: subredditName,
            name: username,
            permissions: permissions ? formatModeratorPermissions(permissions) : undefined,
            ...optionalFields,
        }, metadata);
        if (response.json?.errors?.length) {
            throw new Error(response.json.errors.join('\n'));
        }
    }
    /** @internal */
    static async removeRelationship(options, metadata) {
        const client = Devvit.redditAPIPlugins.Users;
        await client.Unfriend({
            type: options.type,
            subreddit: options.subredditName,
            name: options.username,
        }, metadata);
    }
    /** @internal */
    static async setModeratorPermissions(username, subredditName, permissions, metadata) {
        const client = Devvit.redditAPIPlugins.Users;
        const response = await client.SetPermissions({
            subreddit: subredditName,
            name: username,
            type: 'moderator',
            permissions: formatModeratorPermissions(permissions),
        }, metadata);
        if (response.json?.errors?.length) {
            throw new Error(response.json.errors.join('\n'));
        }
    }
    /** @internal */
    static async getSnoovatarUrl(username, metadata) {
        const operationName = 'GetSnoovatarUrlByName';
        const persistedQueryHash = 'c47fd42345af268616d2d8904b56856acdc05cf61d3650380f539ad7d596ac0c';
        // Legacy GQL query. Do not copy this pattern.
        // eslint-disable-next-line no-restricted-properties
        const response = await GraphQL.query(operationName, persistedQueryHash, { username }, metadata);
        return response.data?.redditorInfoByName?.snoovatarIcon?.url;
    }
    /** @internal */
    static getOverview(options, metadata) {
        const client = Devvit.redditAPIPlugins.Users;
        return new Listing({
            hasMore: true,
            before: options.before,
            after: options.after,
            pageSize: options.pageSize,
            limit: options.limit,
            async fetch(fetchOptions) {
                const response = await client.UserWhere({
                    username: options.username,
                    where: 'overview',
                    ...fetchOptions,
                }, metadata);
                return listingProtosToPostsOrComments(response, metadata);
            },
        });
    }
    /** @internal */
    static async getUserKarmaFromCurrentSubreddit(username, metadata) {
        assertNonNull(metadata);
        const subredditId = metadata?.[Header.Subreddit]?.values[0];
        if (!subredditId) {
            throw new Error('Subreddit ID is missing from metadata');
        }
        return await Devvit.redditAPIPlugins.Users.GetUserKarmaForSubreddit({
            username: username,
            subredditId,
        });
    }
    /** @internal */
    static async getTrophies(username, metadata) {
        const response = await Devvit.redditAPIPlugins.Users.UserTrophies({ username }, metadata);
        return userTrophiesResponseToTrophies(response);
    }
}
_User_id = new WeakMap(), _User_username = new WeakMap(), _User_createdAt = new WeakMap(), _User_linkKarma = new WeakMap(), _User_commentKarma = new WeakMap(), _User_nsfw = new WeakMap(), _User_isAdmin = new WeakMap(), _User_isModerator = new WeakMap(), _User_isGold = new WeakMap(), _User_modPermissionsBySubreddit = new WeakMap(), _User_showNsfw = new WeakMap(), _User_url = new WeakMap(), _User_permalink = new WeakMap(), _User_hasVerifiedEmail = new WeakMap(), _User_displayName = new WeakMap(), _User_about = new WeakMap(), _User_metadata = new WeakMap();
function listingProtosToPostsOrComments(listingProto, metadata) {
    if (!listingProto.data?.children) {
        throw new Error('Listing response is missing children');
    }
    const children = listingProto.data.children.map((child) => {
        if (child.kind === 't3') {
            return new Post(child.data, metadata);
        }
        else if (child.kind === 't1') {
            return new Comment(child.data, metadata);
        }
        throw new Error(`Type ${child.kind} is not supported`);
    });
    return {
        children: children,
        before: listingProto.data.before,
        after: listingProto.data.after,
    };
}
async function listingProtosToUsers(listingProto, subredditName, metadata) {
    const client = Devvit.redditAPIPlugins.Users;
    if (!listingProto.data?.children) {
        throw new Error('Listing response is missing children');
    }
    const userIds = listingProto.data.children.map((child) => {
        assertNonNull(child.data?.id, 'User id is still from listing data');
        return child.data.id;
    });
    // break the ids into chunks since they're passed over a query parameter
    const chunkSize = 100;
    const userIdChunks = [];
    for (let i = 0; i < userIds.length; i += chunkSize) {
        userIdChunks.push(userIds.slice(i, i + chunkSize));
    }
    // perform the requests
    const usersMapResponses = await Promise.all(userIdChunks.map((userIds) => client.UserDataByAccountIds({
        ids: userIds.join(','),
    }, metadata)));
    // join the responses back into a single map of user data
    const userDataById = usersMapResponses.reduce((allUsers, response) => ({ ...allUsers, ...response.users }), {});
    const children = listingProto.data.children.map((child) => {
        const id = child.data?.id;
        assertNonNull(id, 'User id is missing from listing');
        const userData = userDataById[id];
        // Casting to unknown because Typescript assumes that userData is always defined
        // because of how we defined the UserDataByAccountIdsResponse_UserAccountData protobuf.
        assertNonNull(userData, 'User data is missing from response');
        return new User({
            id,
            name: userData.name,
            linkKarma: userData.linkKarma,
            commentKarma: userData.commentKarma,
            createdUtc: userData.createdUtc,
            over18: userData.profileOver18,
            snoovatarSize: [],
            modPermissions: {
                [subredditName]: child.data?.modPermissions ?? [],
            },
        }, metadata);
    });
    return {
        children,
        before: listingProto.data.before,
        after: listingProto.data.after,
    };
}
/** @internal */
function userTrophiesResponseToTrophies(response) {
    if (!response.data?.trophies) {
        return [];
    }
    return response.data.trophies.map((t) => {
        if (!t.data) {
            return {};
        }
        const trophy = {
            id: t.data.id != null && t.data.id !== ''
                ? asT6ID(isT6ID(t.data.id) ? t.data.id : `t6_${t.data.id}`)
                : undefined,
            name: t.data.name,
            description: t.data.description,
            iconUrl: t.data.icon70,
            iconSmallUrl: t.data.icon40,
            grantedAt: t.data.grantedAt,
            awardId: t.data.awardId != null && t.data.awardId !== ''
                ? asT6ID(isT6ID(t.data.awardId) ? t.data.awardId : `t6_${t.data.awardId}`)
                : undefined,
            url: t.data.url,
        };
        return trophy;
    });
}
/** @internal */
async function getUsernameById(id, metadata) {
    const client = Devvit.redditAPIPlugins.Users;
    const response = await client.UserDataByAccountIds({ ids: id }, metadata);
    return response?.users?.[id]?.name;
}
/** @internal */
export async function getCurrentUsernameFromMetadata(metadata) {
    assertNonNull(metadata);
    const username = metadata?.[Header.Username]?.values[0];
    if (username) {
        return username;
    }
    const userId = metadata?.[Header.User]?.values[0];
    if (!userId) {
        return undefined;
    }
    return getUsernameById(userId, metadata);
}
