// @devvit/protos/json/devvit/plugin/redditapi/common/common_msg.js
/**
 * #common_msg.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
/**
 * Label that can be applied to a highlighted post in a subreddit.
 * This enum should map to the equivalent in GraphQL: https://github.snooguts.net/reddit/reddit-service-graphql/blob/master/graphql-py/graphql_api/lib/community_eng/types/highlighted_post.py#L18
 */
export var HighlightedPostLabel;
(function (HighlightedPostLabel) {
    /** HIGHLIGHTED_POST_LABEL_UNSPECIFIED - Post is highlighted without a label. */
    HighlightedPostLabel[HighlightedPostLabel["HIGHLIGHTED_POST_LABEL_UNSPECIFIED"] = 0] = "HIGHLIGHTED_POST_LABEL_UNSPECIFIED";
    /** ANNOUNCEMENT - Post is highlighted as an announcement. */
    HighlightedPostLabel[HighlightedPostLabel["ANNOUNCEMENT"] = 1] = "ANNOUNCEMENT";
    /** EVENT - Post is highlighted as an event. */
    HighlightedPostLabel[HighlightedPostLabel["EVENT"] = 2] = "EVENT";
    /** MEGATHREAD - Post is highlighted as a megathread. */
    HighlightedPostLabel[HighlightedPostLabel["MEGATHREAD"] = 3] = "MEGATHREAD";
    /**
     * SHOW_POST_FLAIR - Post is highlighted with its flair as the label.
     * If the post doesn't have flair, this is the same as UNSPECIFIED
     */
    HighlightedPostLabel[HighlightedPostLabel["SHOW_POST_FLAIR"] = 4] = "SHOW_POST_FLAIR";
    HighlightedPostLabel[HighlightedPostLabel["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(HighlightedPostLabel || (HighlightedPostLabel = {}));
/** Indicates who the request should be made on behalf of. Currently, only 'APP' or 'USER' types are supported. */
export var RunAs;
(function (RunAs) {
    RunAs[RunAs["APP"] = 0] = "APP";
    RunAs[RunAs["USER"] = 1] = "USER";
    RunAs[RunAs["UNSPECIFIED"] = 2] = "UNSPECIFIED";
    RunAs[RunAs["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(RunAs || (RunAs = {}));
export var GalleryMediaStatus;
(function (GalleryMediaStatus) {
    GalleryMediaStatus[GalleryMediaStatus["GALLERY_MEDIA_STATUS_UNSPECIFIED"] = 0] = "GALLERY_MEDIA_STATUS_UNSPECIFIED";
    GalleryMediaStatus[GalleryMediaStatus["GALLERY_MEDIA_STATUS_VALID"] = 1] = "GALLERY_MEDIA_STATUS_VALID";
    GalleryMediaStatus[GalleryMediaStatus["GALLERY_MEDIA_STATUS_FAILED"] = 2] = "GALLERY_MEDIA_STATUS_FAILED";
    GalleryMediaStatus[GalleryMediaStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(GalleryMediaStatus || (GalleryMediaStatus = {}));
/** For methods with a request body, this specifies how the request body should be serialized. */
export var ApiClientConfig_BodyType;
(function (ApiClientConfig_BodyType) {
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["NONE"] = 0] = "NONE";
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["JSON_CAMEL"] = 1] = "JSON_CAMEL";
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["JSON_SNAKE"] = 2] = "JSON_SNAKE";
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["FORM_CAMEL"] = 3] = "FORM_CAMEL";
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["FORM_SNAKE"] = 4] = "FORM_SNAKE";
    ApiClientConfig_BodyType[ApiClientConfig_BodyType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ApiClientConfig_BodyType || (ApiClientConfig_BodyType = {}));
