// @devvit/protos/types/devvit/dev_portal/admin_portal.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { InstallationUpgradeRequest, FullInstallationInfo, InstallationUpgradeManyRequest, InstallationUpgradeManyResponse, } from './installation/installation.js';
import { GetPendingRequest, GetResponse, UpdateStatusForAppRequest, GetFetchDomainRequestsWithGlobalListTypeRequest, } from './fetch_domain_request/fetch_domain_request.js';
import { Empty } from '../../google/protobuf/empty.js';
import { Int32Value } from '../../google/protobuf/wrappers.js';
import { GetFetchDomainsWithStatusRequest, GetFetchDomainsResponse, AddGlobalDomainsRequest, RemoveGlobalDomainRequest, } from './fetch_domain/fetch_domain.js';
import { GetAppBySlugWithDomainInfoRequest, AppWithDomainInfo, GetAppsWithDomainInfoRequest, GetAppsWithDomainInfoResponse, AddDomainsToAppRequest, RemoveDomainFromAppRequest, } from './app/app.js';
import { ListCuratedGamesRequest, ListCuratedGamesResponse, AddCuratedGameRequest, AddCuratedGameResponse, UpdateCuratedGameRankRequest, UpdateCuratedGameRankResponse, RemoveCuratedGameRequest, } from './curated_game/curated_game.js';
import { ListModmailTemplatesRequest, ListModmailTemplatesResponse, CreateModmailTemplateRequest, ModmailTemplate, UpdateModmailTemplateRequest, DeleteModmailTemplateRequest, DeleteModmailTemplateGroupRequest, ReorderModmailTemplatesRequest, GetModmailConfigRequest, ModmailConfig, UpdateModmailConfigRequest, } from './modmail_template/modmail_template.js';
export var AdminInstallationMethod;
(function (AdminInstallationMethod) {
    AdminInstallationMethod["Upgrade"] = "Upgrade";
    AdminInstallationMethod["UpgradeMany"] = "UpgradeMany";
})(AdminInstallationMethod || (AdminInstallationMethod = {}));
export const AdminInstallationMethodList = [
    AdminInstallationMethod.Upgrade,
    AdminInstallationMethod.UpgradeMany,
];
export function createAdminInstallationServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminInstallation',
        methodList: AdminInstallationMethodList,
        matchRoute: matchAdminInstallationRoute,
    });
}
function matchAdminInstallationRoute(method, events) {
    switch (method) {
        case 'Upgrade':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Upgrade' };
                await events.onMatch(ctx);
                return handleAdminInstallationUpgradeRequest(ctx, service, data, interceptors);
            };
        case 'UpgradeMany':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpgradeMany' };
                await events.onMatch(ctx);
                return handleAdminInstallationUpgradeManyRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminInstallationUpgradeRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminInstallationUpgradeJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminInstallationUpgradeProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminInstallationUpgradeManyRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminInstallationUpgradeManyJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminInstallationUpgradeManyProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminInstallationUpgradeJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = InstallationUpgradeRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Upgrade(ctx, inputReq);
        });
    }
    else {
        response = await service.Upgrade(ctx, request);
    }
    return JSON.stringify(FullInstallationInfo.toJSON(response));
}
async function handleAdminInstallationUpgradeManyJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = InstallationUpgradeManyRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpgradeMany(ctx, inputReq);
        });
    }
    else {
        response = await service.UpgradeMany(ctx, request);
    }
    return JSON.stringify(InstallationUpgradeManyResponse.toJSON(response));
}
async function handleAdminInstallationUpgradeProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = InstallationUpgradeRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Upgrade(ctx, inputReq);
        });
    }
    else {
        response = await service.Upgrade(ctx, request);
    }
    return Buffer.from(FullInstallationInfo.encode(response).finish());
}
async function handleAdminInstallationUpgradeManyProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = InstallationUpgradeManyRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpgradeMany(ctx, inputReq);
        });
    }
    else {
        response = await service.UpgradeMany(ctx, request);
    }
    return Buffer.from(InstallationUpgradeManyResponse.encode(response).finish());
}
export var AdminFetchDomainRequestMethod;
(function (AdminFetchDomainRequestMethod) {
    AdminFetchDomainRequestMethod["GetPending"] = "GetPending";
    AdminFetchDomainRequestMethod["GetPendingRequestsCount"] = "GetPendingRequestsCount";
    AdminFetchDomainRequestMethod["UpdateStatusForApp"] = "UpdateStatusForApp";
    AdminFetchDomainRequestMethod["GetFetchDomainRequestsWithGlobalListType"] = "GetFetchDomainRequestsWithGlobalListType";
})(AdminFetchDomainRequestMethod || (AdminFetchDomainRequestMethod = {}));
export const AdminFetchDomainRequestMethodList = [
    AdminFetchDomainRequestMethod.GetPending,
    AdminFetchDomainRequestMethod.GetPendingRequestsCount,
    AdminFetchDomainRequestMethod.UpdateStatusForApp,
    AdminFetchDomainRequestMethod.GetFetchDomainRequestsWithGlobalListType,
];
export function createAdminFetchDomainRequestServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminFetchDomainRequest',
        methodList: AdminFetchDomainRequestMethodList,
        matchRoute: matchAdminFetchDomainRequestRoute,
    });
}
function matchAdminFetchDomainRequestRoute(method, events) {
    switch (method) {
        case 'GetPending':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetPending' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainRequestGetPendingRequest(ctx, service, data, interceptors);
            };
        case 'GetPendingRequestsCount':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetPendingRequestsCount' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainRequestGetPendingRequestsCountRequest(ctx, service, data, interceptors);
            };
        case 'UpdateStatusForApp':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateStatusForApp' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainRequestUpdateStatusForAppRequest(ctx, service, data, interceptors);
            };
        case 'GetFetchDomainRequestsWithGlobalListType':
            return async (ctx, service, data, interceptors) => {
                ctx = {
                    ...ctx,
                    methodName: 'GetFetchDomainRequestsWithGlobalListType',
                };
                await events.onMatch(ctx);
                return handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainRequestGetPendingRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainRequestGetPendingJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainRequestGetPendingProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainRequestGetPendingRequestsCountRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainRequestGetPendingRequestsCountJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainRequestGetPendingRequestsCountProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainRequestUpdateStatusForAppRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainRequestUpdateStatusForAppJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainRequestUpdateStatusForAppProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminFetchDomainRequestGetPendingJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetPendingRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetPending(ctx, inputReq);
        });
    }
    else {
        response = await service.GetPending(ctx, request);
    }
    return JSON.stringify(GetResponse.toJSON(response));
}
async function handleAdminFetchDomainRequestGetPendingRequestsCountJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Empty.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetPendingRequestsCount(ctx, inputReq);
        });
    }
    else {
        response = await service.GetPendingRequestsCount(ctx, request);
    }
    return JSON.stringify(Int32Value.toJSON(response));
}
async function handleAdminFetchDomainRequestUpdateStatusForAppJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateStatusForAppRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateStatusForApp(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateStatusForApp(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetFetchDomainRequestsWithGlobalListTypeRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetFetchDomainRequestsWithGlobalListType(ctx, inputReq);
        });
    }
    else {
        response = await service.GetFetchDomainRequestsWithGlobalListType(ctx, request);
    }
    return JSON.stringify(GetResponse.toJSON(response));
}
async function handleAdminFetchDomainRequestGetPendingProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetPendingRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetPending(ctx, inputReq);
        });
    }
    else {
        response = await service.GetPending(ctx, request);
    }
    return Buffer.from(GetResponse.encode(response).finish());
}
async function handleAdminFetchDomainRequestGetPendingRequestsCountProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Empty.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetPendingRequestsCount(ctx, inputReq);
        });
    }
    else {
        response = await service.GetPendingRequestsCount(ctx, request);
    }
    return Buffer.from(Int32Value.encode(response).finish());
}
async function handleAdminFetchDomainRequestUpdateStatusForAppProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateStatusForAppRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateStatusForApp(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateStatusForApp(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminFetchDomainRequestGetFetchDomainRequestsWithGlobalListTypeProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetFetchDomainRequestsWithGlobalListTypeRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetFetchDomainRequestsWithGlobalListType(ctx, inputReq);
        });
    }
    else {
        response = await service.GetFetchDomainRequestsWithGlobalListType(ctx, request);
    }
    return Buffer.from(GetResponse.encode(response).finish());
}
export var AdminFetchDomainMethod;
(function (AdminFetchDomainMethod) {
    AdminFetchDomainMethod["GetFetchDomainsWithStatus"] = "GetFetchDomainsWithStatus";
    AdminFetchDomainMethod["AddGlobalDomains"] = "AddGlobalDomains";
    AdminFetchDomainMethod["RemoveGlobalDomain"] = "RemoveGlobalDomain";
})(AdminFetchDomainMethod || (AdminFetchDomainMethod = {}));
export const AdminFetchDomainMethodList = [
    AdminFetchDomainMethod.GetFetchDomainsWithStatus,
    AdminFetchDomainMethod.AddGlobalDomains,
    AdminFetchDomainMethod.RemoveGlobalDomain,
];
export function createAdminFetchDomainServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminFetchDomain',
        methodList: AdminFetchDomainMethodList,
        matchRoute: matchAdminFetchDomainRoute,
    });
}
function matchAdminFetchDomainRoute(method, events) {
    switch (method) {
        case 'GetFetchDomainsWithStatus':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetFetchDomainsWithStatus' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainGetFetchDomainsWithStatusRequest(ctx, service, data, interceptors);
            };
        case 'AddGlobalDomains':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'AddGlobalDomains' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainAddGlobalDomainsRequest(ctx, service, data, interceptors);
            };
        case 'RemoveGlobalDomain':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'RemoveGlobalDomain' };
                await events.onMatch(ctx);
                return handleAdminFetchDomainRemoveGlobalDomainRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainGetFetchDomainsWithStatusRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainGetFetchDomainsWithStatusJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainGetFetchDomainsWithStatusProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainAddGlobalDomainsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainAddGlobalDomainsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainAddGlobalDomainsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminFetchDomainRemoveGlobalDomainRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminFetchDomainRemoveGlobalDomainJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminFetchDomainRemoveGlobalDomainProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminFetchDomainGetFetchDomainsWithStatusJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetFetchDomainsWithStatusRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetFetchDomainsWithStatus(ctx, inputReq);
        });
    }
    else {
        response = await service.GetFetchDomainsWithStatus(ctx, request);
    }
    return JSON.stringify(GetFetchDomainsResponse.toJSON(response));
}
async function handleAdminFetchDomainAddGlobalDomainsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = AddGlobalDomainsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddGlobalDomains(ctx, inputReq);
        });
    }
    else {
        response = await service.AddGlobalDomains(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminFetchDomainRemoveGlobalDomainJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = RemoveGlobalDomainRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveGlobalDomain(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveGlobalDomain(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminFetchDomainGetFetchDomainsWithStatusProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetFetchDomainsWithStatusRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetFetchDomainsWithStatus(ctx, inputReq);
        });
    }
    else {
        response = await service.GetFetchDomainsWithStatus(ctx, request);
    }
    return Buffer.from(GetFetchDomainsResponse.encode(response).finish());
}
async function handleAdminFetchDomainAddGlobalDomainsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = AddGlobalDomainsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddGlobalDomains(ctx, inputReq);
        });
    }
    else {
        response = await service.AddGlobalDomains(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminFetchDomainRemoveGlobalDomainProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = RemoveGlobalDomainRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveGlobalDomain(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveGlobalDomain(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
export var AdminAppMethod;
(function (AdminAppMethod) {
    AdminAppMethod["GetAppBySlugWithDomainInfo"] = "GetAppBySlugWithDomainInfo";
    AdminAppMethod["GetAppsWithDomainInfo"] = "GetAppsWithDomainInfo";
    AdminAppMethod["AddDomainsToApp"] = "AddDomainsToApp";
    AdminAppMethod["RemoveDomainFromApp"] = "RemoveDomainFromApp";
})(AdminAppMethod || (AdminAppMethod = {}));
export const AdminAppMethodList = [
    AdminAppMethod.GetAppBySlugWithDomainInfo,
    AdminAppMethod.GetAppsWithDomainInfo,
    AdminAppMethod.AddDomainsToApp,
    AdminAppMethod.RemoveDomainFromApp,
];
export function createAdminAppServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminApp',
        methodList: AdminAppMethodList,
        matchRoute: matchAdminAppRoute,
    });
}
function matchAdminAppRoute(method, events) {
    switch (method) {
        case 'GetAppBySlugWithDomainInfo':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetAppBySlugWithDomainInfo' };
                await events.onMatch(ctx);
                return handleAdminAppGetAppBySlugWithDomainInfoRequest(ctx, service, data, interceptors);
            };
        case 'GetAppsWithDomainInfo':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetAppsWithDomainInfo' };
                await events.onMatch(ctx);
                return handleAdminAppGetAppsWithDomainInfoRequest(ctx, service, data, interceptors);
            };
        case 'AddDomainsToApp':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'AddDomainsToApp' };
                await events.onMatch(ctx);
                return handleAdminAppAddDomainsToAppRequest(ctx, service, data, interceptors);
            };
        case 'RemoveDomainFromApp':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'RemoveDomainFromApp' };
                await events.onMatch(ctx);
                return handleAdminAppRemoveDomainFromAppRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminAppGetAppBySlugWithDomainInfoRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminAppGetAppBySlugWithDomainInfoJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminAppGetAppBySlugWithDomainInfoProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminAppGetAppsWithDomainInfoRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminAppGetAppsWithDomainInfoJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminAppGetAppsWithDomainInfoProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminAppAddDomainsToAppRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminAppAddDomainsToAppJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminAppAddDomainsToAppProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminAppRemoveDomainFromAppRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminAppRemoveDomainFromAppJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminAppRemoveDomainFromAppProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminAppGetAppBySlugWithDomainInfoJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetAppBySlugWithDomainInfoRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppBySlugWithDomainInfo(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppBySlugWithDomainInfo(ctx, request);
    }
    return JSON.stringify(AppWithDomainInfo.toJSON(response));
}
async function handleAdminAppGetAppsWithDomainInfoJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetAppsWithDomainInfoRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppsWithDomainInfo(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppsWithDomainInfo(ctx, request);
    }
    return JSON.stringify(GetAppsWithDomainInfoResponse.toJSON(response));
}
async function handleAdminAppAddDomainsToAppJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = AddDomainsToAppRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddDomainsToApp(ctx, inputReq);
        });
    }
    else {
        response = await service.AddDomainsToApp(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminAppRemoveDomainFromAppJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = RemoveDomainFromAppRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveDomainFromApp(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveDomainFromApp(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminAppGetAppBySlugWithDomainInfoProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetAppBySlugWithDomainInfoRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppBySlugWithDomainInfo(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppBySlugWithDomainInfo(ctx, request);
    }
    return Buffer.from(AppWithDomainInfo.encode(response).finish());
}
async function handleAdminAppGetAppsWithDomainInfoProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetAppsWithDomainInfoRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppsWithDomainInfo(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppsWithDomainInfo(ctx, request);
    }
    return Buffer.from(GetAppsWithDomainInfoResponse.encode(response).finish());
}
async function handleAdminAppAddDomainsToAppProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = AddDomainsToAppRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddDomainsToApp(ctx, inputReq);
        });
    }
    else {
        response = await service.AddDomainsToApp(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminAppRemoveDomainFromAppProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = RemoveDomainFromAppRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveDomainFromApp(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveDomainFromApp(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
export var AdminCuratedGameMethod;
(function (AdminCuratedGameMethod) {
    AdminCuratedGameMethod["ListGames"] = "ListGames";
    AdminCuratedGameMethod["AddGame"] = "AddGame";
    AdminCuratedGameMethod["UpdateGameRank"] = "UpdateGameRank";
    AdminCuratedGameMethod["RemoveGame"] = "RemoveGame";
})(AdminCuratedGameMethod || (AdminCuratedGameMethod = {}));
export const AdminCuratedGameMethodList = [
    AdminCuratedGameMethod.ListGames,
    AdminCuratedGameMethod.AddGame,
    AdminCuratedGameMethod.UpdateGameRank,
    AdminCuratedGameMethod.RemoveGame,
];
export function createAdminCuratedGameServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminCuratedGame',
        methodList: AdminCuratedGameMethodList,
        matchRoute: matchAdminCuratedGameRoute,
    });
}
function matchAdminCuratedGameRoute(method, events) {
    switch (method) {
        case 'ListGames':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'ListGames' };
                await events.onMatch(ctx);
                return handleAdminCuratedGameListGamesRequest(ctx, service, data, interceptors);
            };
        case 'AddGame':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'AddGame' };
                await events.onMatch(ctx);
                return handleAdminCuratedGameAddGameRequest(ctx, service, data, interceptors);
            };
        case 'UpdateGameRank':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateGameRank' };
                await events.onMatch(ctx);
                return handleAdminCuratedGameUpdateGameRankRequest(ctx, service, data, interceptors);
            };
        case 'RemoveGame':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'RemoveGame' };
                await events.onMatch(ctx);
                return handleAdminCuratedGameRemoveGameRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminCuratedGameListGamesRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminCuratedGameListGamesJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminCuratedGameListGamesProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminCuratedGameAddGameRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminCuratedGameAddGameJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminCuratedGameAddGameProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminCuratedGameUpdateGameRankRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminCuratedGameUpdateGameRankJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminCuratedGameUpdateGameRankProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminCuratedGameRemoveGameRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminCuratedGameRemoveGameJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminCuratedGameRemoveGameProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminCuratedGameListGamesJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ListCuratedGamesRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ListGames(ctx, inputReq);
        });
    }
    else {
        response = await service.ListGames(ctx, request);
    }
    return JSON.stringify(ListCuratedGamesResponse.toJSON(response));
}
async function handleAdminCuratedGameAddGameJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = AddCuratedGameRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddGame(ctx, inputReq);
        });
    }
    else {
        response = await service.AddGame(ctx, request);
    }
    return JSON.stringify(AddCuratedGameResponse.toJSON(response));
}
async function handleAdminCuratedGameUpdateGameRankJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateCuratedGameRankRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateGameRank(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateGameRank(ctx, request);
    }
    return JSON.stringify(UpdateCuratedGameRankResponse.toJSON(response));
}
async function handleAdminCuratedGameRemoveGameJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = RemoveCuratedGameRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveGame(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveGame(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminCuratedGameListGamesProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ListCuratedGamesRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ListGames(ctx, inputReq);
        });
    }
    else {
        response = await service.ListGames(ctx, request);
    }
    return Buffer.from(ListCuratedGamesResponse.encode(response).finish());
}
async function handleAdminCuratedGameAddGameProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = AddCuratedGameRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddGame(ctx, inputReq);
        });
    }
    else {
        response = await service.AddGame(ctx, request);
    }
    return Buffer.from(AddCuratedGameResponse.encode(response).finish());
}
async function handleAdminCuratedGameUpdateGameRankProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateCuratedGameRankRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateGameRank(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateGameRank(ctx, request);
    }
    return Buffer.from(UpdateCuratedGameRankResponse.encode(response).finish());
}
async function handleAdminCuratedGameRemoveGameProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = RemoveCuratedGameRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemoveGame(ctx, inputReq);
        });
    }
    else {
        response = await service.RemoveGame(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
export var AdminModmailTemplateMethod;
(function (AdminModmailTemplateMethod) {
    AdminModmailTemplateMethod["List"] = "List";
    AdminModmailTemplateMethod["Create"] = "Create";
    AdminModmailTemplateMethod["Update"] = "Update";
    AdminModmailTemplateMethod["Delete"] = "Delete";
    AdminModmailTemplateMethod["DeleteGroup"] = "DeleteGroup";
    AdminModmailTemplateMethod["Reorder"] = "Reorder";
    AdminModmailTemplateMethod["GetConfig"] = "GetConfig";
    AdminModmailTemplateMethod["UpdateConfig"] = "UpdateConfig";
})(AdminModmailTemplateMethod || (AdminModmailTemplateMethod = {}));
export const AdminModmailTemplateMethodList = [
    AdminModmailTemplateMethod.List,
    AdminModmailTemplateMethod.Create,
    AdminModmailTemplateMethod.Update,
    AdminModmailTemplateMethod.Delete,
    AdminModmailTemplateMethod.DeleteGroup,
    AdminModmailTemplateMethod.Reorder,
    AdminModmailTemplateMethod.GetConfig,
    AdminModmailTemplateMethod.UpdateConfig,
];
export function createAdminModmailTemplateServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.dev_portal',
        serviceName: 'AdminModmailTemplate',
        methodList: AdminModmailTemplateMethodList,
        matchRoute: matchAdminModmailTemplateRoute,
    });
}
function matchAdminModmailTemplateRoute(method, events) {
    switch (method) {
        case 'List':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'List' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateListRequest(ctx, service, data, interceptors);
            };
        case 'Create':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Create' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateCreateRequest(ctx, service, data, interceptors);
            };
        case 'Update':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Update' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateUpdateRequest(ctx, service, data, interceptors);
            };
        case 'Delete':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Delete' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateDeleteRequest(ctx, service, data, interceptors);
            };
        case 'DeleteGroup':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'DeleteGroup' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateDeleteGroupRequest(ctx, service, data, interceptors);
            };
        case 'Reorder':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Reorder' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateReorderRequest(ctx, service, data, interceptors);
            };
        case 'GetConfig':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetConfig' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateGetConfigRequest(ctx, service, data, interceptors);
            };
        case 'UpdateConfig':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateConfig' };
                await events.onMatch(ctx);
                return handleAdminModmailTemplateUpdateConfigRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateListRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateListJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateListProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateCreateRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateCreateJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateCreateProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateUpdateRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateUpdateJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateUpdateProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateDeleteRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateDeleteJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateDeleteProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateDeleteGroupRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateDeleteGroupJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateDeleteGroupProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateReorderRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateReorderJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateReorderProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateGetConfigRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateGetConfigJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateGetConfigProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAdminModmailTemplateUpdateConfigRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAdminModmailTemplateUpdateConfigJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAdminModmailTemplateUpdateConfigProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAdminModmailTemplateListJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ListModmailTemplatesRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.List(ctx, inputReq);
        });
    }
    else {
        response = await service.List(ctx, request);
    }
    return JSON.stringify(ListModmailTemplatesResponse.toJSON(response));
}
async function handleAdminModmailTemplateCreateJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = CreateModmailTemplateRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Create(ctx, inputReq);
        });
    }
    else {
        response = await service.Create(ctx, request);
    }
    return JSON.stringify(ModmailTemplate.toJSON(response));
}
async function handleAdminModmailTemplateUpdateJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateModmailTemplateRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Update(ctx, inputReq);
        });
    }
    else {
        response = await service.Update(ctx, request);
    }
    return JSON.stringify(ModmailTemplate.toJSON(response));
}
async function handleAdminModmailTemplateDeleteJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = DeleteModmailTemplateRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Delete(ctx, inputReq);
        });
    }
    else {
        response = await service.Delete(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminModmailTemplateDeleteGroupJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = DeleteModmailTemplateGroupRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.DeleteGroup(ctx, inputReq);
        });
    }
    else {
        response = await service.DeleteGroup(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminModmailTemplateReorderJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ReorderModmailTemplatesRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Reorder(ctx, inputReq);
        });
    }
    else {
        response = await service.Reorder(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleAdminModmailTemplateGetConfigJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetModmailConfigRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetConfig(ctx, inputReq);
        });
    }
    else {
        response = await service.GetConfig(ctx, request);
    }
    return JSON.stringify(ModmailConfig.toJSON(response));
}
async function handleAdminModmailTemplateUpdateConfigJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateModmailConfigRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateConfig(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateConfig(ctx, request);
    }
    return JSON.stringify(ModmailConfig.toJSON(response));
}
async function handleAdminModmailTemplateListProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ListModmailTemplatesRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.List(ctx, inputReq);
        });
    }
    else {
        response = await service.List(ctx, request);
    }
    return Buffer.from(ListModmailTemplatesResponse.encode(response).finish());
}
async function handleAdminModmailTemplateCreateProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = CreateModmailTemplateRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Create(ctx, inputReq);
        });
    }
    else {
        response = await service.Create(ctx, request);
    }
    return Buffer.from(ModmailTemplate.encode(response).finish());
}
async function handleAdminModmailTemplateUpdateProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateModmailTemplateRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Update(ctx, inputReq);
        });
    }
    else {
        response = await service.Update(ctx, request);
    }
    return Buffer.from(ModmailTemplate.encode(response).finish());
}
async function handleAdminModmailTemplateDeleteProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = DeleteModmailTemplateRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Delete(ctx, inputReq);
        });
    }
    else {
        response = await service.Delete(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminModmailTemplateDeleteGroupProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = DeleteModmailTemplateGroupRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.DeleteGroup(ctx, inputReq);
        });
    }
    else {
        response = await service.DeleteGroup(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminModmailTemplateReorderProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ReorderModmailTemplatesRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Reorder(ctx, inputReq);
        });
    }
    else {
        response = await service.Reorder(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handleAdminModmailTemplateGetConfigProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetModmailConfigRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetConfig(ctx, inputReq);
        });
    }
    else {
        response = await service.GetConfig(ctx, request);
    }
    return Buffer.from(ModmailConfig.encode(response).finish());
}
async function handleAdminModmailTemplateUpdateConfigProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateModmailConfigRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateConfig(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateConfig(ctx, request);
    }
    return Buffer.from(ModmailConfig.encode(response).finish());
}
