// @devvit/protos/types/devvit/dev_portal/fetch_domain/fetch_domain.js
/**
 * #fetch_domain.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { Timestamp } from "../../../google/protobuf/timestamp.js";
import { messageTypeRegistry } from "../../../typeRegistry.js";
import { globalListTypeFromJSON, globalListTypeToJSON, } from "../fetch_domain_request/fetch_domain_request.js";
/** Represents a domain's global status. UNDECIDED means the domain is not globally decided, ie only applies at the app level. */
export var FetchDomainGlobalStatus;
(function (FetchDomainGlobalStatus) {
    FetchDomainGlobalStatus[FetchDomainGlobalStatus["UNDECIDED"] = 0] = "UNDECIDED";
    FetchDomainGlobalStatus[FetchDomainGlobalStatus["ALLOWED"] = 1] = "ALLOWED";
    FetchDomainGlobalStatus[FetchDomainGlobalStatus["DENIED"] = 2] = "DENIED";
    FetchDomainGlobalStatus[FetchDomainGlobalStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(FetchDomainGlobalStatus || (FetchDomainGlobalStatus = {}));
export function fetchDomainGlobalStatusFromJSON(object) {
    switch (object) {
        case 0:
        case "UNDECIDED":
            return FetchDomainGlobalStatus.UNDECIDED;
        case 1:
        case "ALLOWED":
            return FetchDomainGlobalStatus.ALLOWED;
        case 2:
        case "DENIED":
            return FetchDomainGlobalStatus.DENIED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return FetchDomainGlobalStatus.UNRECOGNIZED;
    }
}
export function fetchDomainGlobalStatusToJSON(object) {
    switch (object) {
        case FetchDomainGlobalStatus.UNDECIDED:
            return 0;
        case FetchDomainGlobalStatus.ALLOWED:
            return 1;
        case FetchDomainGlobalStatus.DENIED:
            return 2;
        case FetchDomainGlobalStatus.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseFetchDomain() {
    return { domain: "", globalStatus: 0, updatedAt: undefined };
}
export const FetchDomain = {
    $type: "devvit.dev_portal.fetch_domain.FetchDomain",
    encode(message, writer = _m0.Writer.create()) {
        if (message.domain !== "") {
            writer.uint32(10).string(message.domain);
        }
        if (message.globalStatus !== 0) {
            writer.uint32(16).int32(message.globalStatus);
        }
        if (message.updatedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.updatedAt), writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFetchDomain();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.globalStatus = reader.int32();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.updatedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            domain: isSet(object.domain) ? globalThis.String(object.domain) : "",
            globalStatus: isSet(object.globalStatus) ? fetchDomainGlobalStatusFromJSON(object.globalStatus) : 0,
            updatedAt: isSet(object.updatedAt) ? fromJsonTimestamp(object.updatedAt) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.domain !== "") {
            obj.domain = message.domain;
        }
        if (message.globalStatus !== 0) {
            obj.globalStatus = fetchDomainGlobalStatusToJSON(message.globalStatus);
        }
        if (message.updatedAt !== undefined) {
            obj.updatedAt = message.updatedAt.toISOString();
        }
        return obj;
    },
    create(base) {
        return FetchDomain.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseFetchDomain();
        message.domain = object.domain ?? "";
        message.globalStatus = object.globalStatus ?? 0;
        message.updatedAt = object.updatedAt ?? undefined;
        return message;
    },
};
messageTypeRegistry.set(FetchDomain.$type, FetchDomain);
function createBaseGetFetchDomainsWithStatusRequest() {
    return { globalStatus: 0, pageNumber: 0, pageSize: 0, domainFilter: undefined };
}
export const GetFetchDomainsWithStatusRequest = {
    $type: "devvit.dev_portal.fetch_domain.GetFetchDomainsWithStatusRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.globalStatus !== 0) {
            writer.uint32(8).int32(message.globalStatus);
        }
        if (message.pageNumber !== 0) {
            writer.uint32(16).int32(message.pageNumber);
        }
        if (message.pageSize !== 0) {
            writer.uint32(24).int32(message.pageSize);
        }
        if (message.domainFilter !== undefined) {
            writer.uint32(34).string(message.domainFilter);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetFetchDomainsWithStatusRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.globalStatus = reader.int32();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.pageNumber = reader.int32();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.pageSize = reader.int32();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.domainFilter = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            globalStatus: isSet(object.globalStatus) ? fetchDomainGlobalStatusFromJSON(object.globalStatus) : 0,
            pageNumber: isSet(object.pageNumber) ? globalThis.Number(object.pageNumber) : 0,
            pageSize: isSet(object.pageSize) ? globalThis.Number(object.pageSize) : 0,
            domainFilter: isSet(object.domainFilter) ? globalThis.String(object.domainFilter) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.globalStatus !== 0) {
            obj.globalStatus = fetchDomainGlobalStatusToJSON(message.globalStatus);
        }
        if (message.pageNumber !== 0) {
            obj.pageNumber = Math.round(message.pageNumber);
        }
        if (message.pageSize !== 0) {
            obj.pageSize = Math.round(message.pageSize);
        }
        if (message.domainFilter !== undefined) {
            obj.domainFilter = message.domainFilter;
        }
        return obj;
    },
    create(base) {
        return GetFetchDomainsWithStatusRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetFetchDomainsWithStatusRequest();
        message.globalStatus = object.globalStatus ?? 0;
        message.pageNumber = object.pageNumber ?? 0;
        message.pageSize = object.pageSize ?? 0;
        message.domainFilter = object.domainFilter ?? undefined;
        return message;
    },
};
messageTypeRegistry.set(GetFetchDomainsWithStatusRequest.$type, GetFetchDomainsWithStatusRequest);
function createBaseGetFetchDomainsResponse() {
    return { fetchDomains: [], hasNextPage: undefined, totalCount: undefined };
}
export const GetFetchDomainsResponse = {
    $type: "devvit.dev_portal.fetch_domain.GetFetchDomainsResponse",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.fetchDomains) {
            FetchDomain.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.hasNextPage !== undefined) {
            writer.uint32(16).bool(message.hasNextPage);
        }
        if (message.totalCount !== undefined) {
            writer.uint32(24).int32(message.totalCount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetFetchDomainsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.fetchDomains.push(FetchDomain.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.hasNextPage = reader.bool();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.totalCount = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            fetchDomains: globalThis.Array.isArray(object?.fetchDomains)
                ? object.fetchDomains.map((e) => FetchDomain.fromJSON(e))
                : [],
            hasNextPage: isSet(object.hasNextPage) ? globalThis.Boolean(object.hasNextPage) : undefined,
            totalCount: isSet(object.totalCount) ? globalThis.Number(object.totalCount) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.fetchDomains?.length) {
            obj.fetchDomains = message.fetchDomains.map((e) => FetchDomain.toJSON(e));
        }
        if (message.hasNextPage !== undefined) {
            obj.hasNextPage = message.hasNextPage;
        }
        if (message.totalCount !== undefined) {
            obj.totalCount = Math.round(message.totalCount);
        }
        return obj;
    },
    create(base) {
        return GetFetchDomainsResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseGetFetchDomainsResponse();
        message.fetchDomains = object.fetchDomains?.map((e) => FetchDomain.fromPartial(e)) || [];
        message.hasNextPage = object.hasNextPage ?? undefined;
        message.totalCount = object.totalCount ?? undefined;
        return message;
    },
};
messageTypeRegistry.set(GetFetchDomainsResponse.$type, GetFetchDomainsResponse);
function createBaseAddGlobalDomainsRequest() {
    return { domains: [], globalStatus: 0 };
}
export const AddGlobalDomainsRequest = {
    $type: "devvit.dev_portal.fetch_domain.AddGlobalDomainsRequest",
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.domains) {
            writer.uint32(10).string(v);
        }
        if (message.globalStatus !== 0) {
            writer.uint32(16).int32(message.globalStatus);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddGlobalDomainsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.domains.push(reader.string());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.globalStatus = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            domains: globalThis.Array.isArray(object?.domains) ? object.domains.map((e) => globalThis.String(e)) : [],
            globalStatus: isSet(object.globalStatus) ? fetchDomainGlobalStatusFromJSON(object.globalStatus) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.domains?.length) {
            obj.domains = message.domains;
        }
        if (message.globalStatus !== 0) {
            obj.globalStatus = fetchDomainGlobalStatusToJSON(message.globalStatus);
        }
        return obj;
    },
    create(base) {
        return AddGlobalDomainsRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAddGlobalDomainsRequest();
        message.domains = object.domains?.map((e) => e) || [];
        message.globalStatus = object.globalStatus ?? 0;
        return message;
    },
};
messageTypeRegistry.set(AddGlobalDomainsRequest.$type, AddGlobalDomainsRequest);
function createBaseRemoveGlobalDomainRequest() {
    return { domain: "", globalListType: 0 };
}
export const RemoveGlobalDomainRequest = {
    $type: "devvit.dev_portal.fetch_domain.RemoveGlobalDomainRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.domain !== "") {
            writer.uint32(10).string(message.domain);
        }
        if (message.globalListType !== 0) {
            writer.uint32(16).int32(message.globalListType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveGlobalDomainRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.globalListType = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            domain: isSet(object.domain) ? globalThis.String(object.domain) : "",
            globalListType: isSet(object.globalListType) ? globalListTypeFromJSON(object.globalListType) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.domain !== "") {
            obj.domain = message.domain;
        }
        if (message.globalListType !== 0) {
            obj.globalListType = globalListTypeToJSON(message.globalListType);
        }
        return obj;
    },
    create(base) {
        return RemoveGlobalDomainRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRemoveGlobalDomainRequest();
        message.domain = object.domain ?? "";
        message.globalListType = object.globalListType ?? 0;
        return message;
    },
};
messageTypeRegistry.set(RemoveGlobalDomainRequest.$type, RemoveGlobalDomainRequest);
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000);
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = (t.seconds || 0) * 1000;
    millis += (t.nanos || 0) / 1000000;
    return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof globalThis.Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new globalThis.Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
