// @devvit/protos/json/devvit/reddit/v2alpha/subredditv2.js
/**
 * #subredditv2.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var SubredditType;
(function (SubredditType) {
    SubredditType[SubredditType["ARCHIVED"] = 0] = "ARCHIVED";
    SubredditType[SubredditType["EMPLOYEES_ONLY"] = 1] = "EMPLOYEES_ONLY";
    SubredditType[SubredditType["GOLD_ONLY"] = 2] = "GOLD_ONLY";
    SubredditType[SubredditType["GOLD_RESTRICTED"] = 3] = "GOLD_RESTRICTED";
    SubredditType[SubredditType["PRIVATE"] = 4] = "PRIVATE";
    SubredditType[SubredditType["PUBLIC"] = 5] = "PUBLIC";
    SubredditType[SubredditType["RESTRICTED"] = 6] = "RESTRICTED";
    SubredditType[SubredditType["USER"] = 7] = "USER";
    SubredditType[SubredditType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SubredditType || (SubredditType = {}));
/** used with data-api */
export var SubredditRating;
(function (SubredditRating) {
    SubredditRating[SubredditRating["UNKNOWN_SUBREDDIT_RATING"] = 0] = "UNKNOWN_SUBREDDIT_RATING";
    /** E - everyone */
    SubredditRating[SubredditRating["E"] = 1] = "E";
    /** M1 - mature 1 */
    SubredditRating[SubredditRating["M1"] = 2] = "M1";
    /** M2 - mature 2 */
    SubredditRating[SubredditRating["M2"] = 3] = "M2";
    /** D - high-risk drug use */
    SubredditRating[SubredditRating["D"] = 4] = "D";
    /** V - violence & gore */
    SubredditRating[SubredditRating["V"] = 5] = "V";
    /** X - sexually explicit */
    SubredditRating[SubredditRating["X"] = 6] = "X";
    SubredditRating[SubredditRating["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SubredditRating || (SubredditRating = {}));
/** creating new enums so that we can have unspecified and unset values */
export var SubredditTypeV2;
(function (SubredditTypeV2) {
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_UNSPECIFIED"] = 0] = "SUBREDDIT_TYPE_UNSPECIFIED";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_UNKNOWN"] = 1] = "SUBREDDIT_TYPE_UNKNOWN";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_ARCHIVED"] = 2] = "SUBREDDIT_TYPE_ARCHIVED";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_EMPLOYEES_ONLY"] = 3] = "SUBREDDIT_TYPE_EMPLOYEES_ONLY";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_GOLD_ONLY"] = 4] = "SUBREDDIT_TYPE_GOLD_ONLY";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_GOLD_RESTRICTED"] = 5] = "SUBREDDIT_TYPE_GOLD_RESTRICTED";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_PRIVATE"] = 6] = "SUBREDDIT_TYPE_PRIVATE";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_PUBLIC"] = 7] = "SUBREDDIT_TYPE_PUBLIC";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_RESTRICTED"] = 8] = "SUBREDDIT_TYPE_RESTRICTED";
    SubredditTypeV2[SubredditTypeV2["SUBREDDIT_TYPE_USER"] = 9] = "SUBREDDIT_TYPE_USER";
    SubredditTypeV2[SubredditTypeV2["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SubredditTypeV2 || (SubredditTypeV2 = {}));
export var SubredditRatingV2;
(function (SubredditRatingV2) {
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_UNSPECIFIED"] = 0] = "SUBREDDIT_RATING_UNSPECIFIED";
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_UNKNOWN"] = 1] = "SUBREDDIT_RATING_UNKNOWN";
    /** SUBREDDIT_RATING_E - everyone */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_E"] = 2] = "SUBREDDIT_RATING_E";
    /** SUBREDDIT_RATING_M1 - mature 1 */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_M1"] = 3] = "SUBREDDIT_RATING_M1";
    /** SUBREDDIT_RATING_M2 - mature 2 */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_M2"] = 4] = "SUBREDDIT_RATING_M2";
    /** SUBREDDIT_RATING_D - high-risk drug use */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_D"] = 5] = "SUBREDDIT_RATING_D";
    /** SUBREDDIT_RATING_V - violence & gore */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_V"] = 6] = "SUBREDDIT_RATING_V";
    /** SUBREDDIT_RATING_X - sexually explicit */
    SubredditRatingV2[SubredditRatingV2["SUBREDDIT_RATING_X"] = 7] = "SUBREDDIT_RATING_X";
    SubredditRatingV2[SubredditRatingV2["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SubredditRatingV2 || (SubredditRatingV2 = {}));
