// @devvit/protos/json/devvit/dev_portal/data_api/admin/audit/audit.js
/**
 * #audit.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var AuditLogEventType;
(function (AuditLogEventType) {
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_CREATE"] = 0] = "DATA_SUBSCRIPTION_CREATE";
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_UPDATE"] = 1] = "DATA_SUBSCRIPTION_UPDATE";
    AuditLogEventType[AuditLogEventType["DATA_SUBSCRIPTION_DELETE"] = 2] = "DATA_SUBSCRIPTION_DELETE";
    AuditLogEventType[AuditLogEventType["DATA_TOKEN_CREATE"] = 3] = "DATA_TOKEN_CREATE";
    AuditLogEventType[AuditLogEventType["DATA_TOKEN_REVOKE"] = 4] = "DATA_TOKEN_REVOKE";
    AuditLogEventType[AuditLogEventType["ORGANIZATION_DELETE"] = 5] = "ORGANIZATION_DELETE";
    AuditLogEventType[AuditLogEventType["ORGANIZATION_UPDATE"] = 6] = "ORGANIZATION_UPDATE";
    AuditLogEventType[AuditLogEventType["ORGANIZATION_CREATE"] = 7] = "ORGANIZATION_CREATE";
    AuditLogEventType[AuditLogEventType["USER_CREATE"] = 8] = "USER_CREATE";
    AuditLogEventType[AuditLogEventType["USER_DELETE"] = 9] = "USER_DELETE";
    AuditLogEventType[AuditLogEventType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AuditLogEventType || (AuditLogEventType = {}));
