// @devvit/protos/json/devvit/gateway/utils/v1alpha/utilities.js
/**
 * #utilities.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var CurationType;
(function (CurationType) {
    CurationType[CurationType["CURATION_TYPE_UNSPECIFIED"] = 0] = "CURATION_TYPE_UNSPECIFIED";
    CurationType[CurationType["CURATION_TYPE_NEW"] = 1] = "CURATION_TYPE_NEW";
    CurationType[CurationType["CURATION_TYPE_POPULAR"] = 2] = "CURATION_TYPE_POPULAR";
    CurationType[CurationType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(CurationType || (CurationType = {}));
/** Purge all cache entries of the specified type */
export var PurgeAllRequest_CacheType;
(function (PurgeAllRequest_CacheType) {
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["UNSPECIFIED"] = 0] = "UNSPECIFIED";
    /** INSTALLATION - devvit-ecosystem cache types */
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["INSTALLATION"] = 1] = "INSTALLATION";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["SUBREDDIT_APP_INSTALLS"] = 2] = "SUBREDDIT_APP_INSTALLS";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP"] = 3] = "APP";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_BY_ID"] = 4] = "APP_BY_ID";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_PUBLISH_REQUEST"] = 5] = "APP_PUBLISH_REQUEST";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_DOMAINS"] = 6] = "APP_DOMAINS";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["GLOBAL_DOMAINS"] = 7] = "GLOBAL_DOMAINS";
    /** POST - devvit-runtime cache types */
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["POST"] = 8] = "POST";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["POST_RENDER"] = 9] = "POST_RENDER";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["POSTS_BY_INSTALLATION"] = 10] = "POSTS_BY_INSTALLATION";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["CONTEXT_ACTIONS"] = 11] = "CONTEXT_ACTIONS";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_PERMISSIONS"] = 12] = "APP_PERMISSIONS";
    /** DRAWER - other cache types */
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["DRAWER"] = 13] = "DRAWER";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["TOP_CUSTOM_POSTS"] = 14] = "TOP_CUSTOM_POSTS";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_HEALTH"] = 15] = "APP_HEALTH";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["APP_WITH_CATEGORIES"] = 16] = "APP_WITH_CATEGORIES";
    PurgeAllRequest_CacheType[PurgeAllRequest_CacheType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(PurgeAllRequest_CacheType || (PurgeAllRequest_CacheType = {}));
