// @devvit/protos/json/devvit/dev_portal/nutrition/nutrition.js
/**
 * #nutrition.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
/**
 * Abstract plugin capabilities required by an app. Plugin usage is bucketed
 * coarsely by Definition.fullName and may change. This enumeration is to power
 * informative permission UI, not make platform security decisions. See the
 * nutritionCategoriesFromBundle() implementation for the current behavior.
 *
 * "Use" and "provides" terminology is loose. An app that _uses_ a scheduler
 * callback OR _provides_ the scheduler plugin is still categorized as
 * "SCHEDULER".
 *
 * See ActorType in packages/protos/src/index.ts. Beware: many of these
 * Defintions should never be Devvit apps.
 */
export var NutritionCategory;
(function (NutritionCategory) {
    /** UNCATEGORIZED - App may use uncategorized permissions. */
    NutritionCategory[NutritionCategory["UNCATEGORIZED"] = 0] = "UNCATEGORIZED";
    /** APP_TRIGGERS - App may execute when installed or upgraded. */
    NutritionCategory[NutritionCategory["APP_TRIGGERS"] = 1] = "APP_TRIGGERS";
    /** ASSETS - App may present predefined audio-visual media. */
    NutritionCategory[NutritionCategory["ASSETS"] = 2] = "ASSETS";
    /** DATA - App may read and write _app_ data to Reddit servers. */
    NutritionCategory[NutritionCategory["DATA"] = 3] = "DATA";
    /** HTTP - App may read and write _any_ data to and from the internet. */
    NutritionCategory[NutritionCategory["HTTP"] = 4] = "HTTP";
    /**
     * MODLOG - App may read and write data to Reddit moderator logs. Does not imply Reddit
     * API.
     * DEPRECATED: The modlog api has been removed.
     */
    NutritionCategory[NutritionCategory["MODLOG"] = 5] = "MODLOG";
    /** REDDIT_API - App may read and write _any_ Reddit data. Does not imply HTTP. */
    NutritionCategory[NutritionCategory["REDDIT_API"] = 6] = "REDDIT_API";
    /**
     * REDDIT_TRIGGERS - App may observe Reddit events such as post creation, comment deletion, and
     * many more. Does not imply REDDIT_API.
     */
    NutritionCategory[NutritionCategory["REDDIT_TRIGGERS"] = 7] = "REDDIT_TRIGGERS";
    /** SCHEDULER - App may schedule itself for later execution. */
    NutritionCategory[NutritionCategory["SCHEDULER"] = 8] = "SCHEDULER";
    /**
     * UI - App may appear in subreddit, post, and comment menu entries, and custom posts.
     * Note: CUSTOM_POST is a subset of UI: if CUSTOM_POST is set then UI is set also.
     */
    NutritionCategory[NutritionCategory["UI"] = 9] = "UI";
    /** MODERATOR - App may appear and act as a moderator on subreddits where it's installed. */
    NutritionCategory[NutritionCategory["MODERATOR"] = 10] = "MODERATOR";
    /** WEBVIEW - App may embed a webview to enable more web functionality in their app. */
    NutritionCategory[NutritionCategory["WEBVIEW"] = 11] = "WEBVIEW";
    /** PAYMENTS - App may sell digital goods or services. */
    NutritionCategory[NutritionCategory["PAYMENTS"] = 12] = "PAYMENTS";
    /**
     * CUSTOM_POST - App can create a custom experience.
     * Note: CUSTOM_POST is a subset of UI: if CUSTOM_POST is set then UI is set also.
     */
    NutritionCategory[NutritionCategory["CUSTOM_POST"] = 13] = "CUSTOM_POST";
    /** USER_ACTIONS - App may invoke Reddit APIs on behalf of the user. */
    NutritionCategory[NutritionCategory["USER_ACTIONS"] = 14] = "USER_ACTIONS";
    /** EXTERNAL_ENDPOINTS - App may be invoked by third-parties outside of Reddit. */
    NutritionCategory[NutritionCategory["EXTERNAL_ENDPOINTS"] = 15] = "EXTERNAL_ENDPOINTS";
    NutritionCategory[NutritionCategory["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(NutritionCategory || (NutritionCategory = {}));
