// @devvit/shared-types/Header.js
/**
 * Metadata header key. Every system header should start with "devvit-".
 *
 * Synchronize to headers.md.
 */
export const Header = Object.freeze({
    Actor: 'devvit-actor',
    App: 'devvit-app',
    AppUser: 'devvit-app-user',
    Canary: 'devvit-canary',
    Comment: 'devvit-comment',
    Context: 'devvit-context',
    Debug: 'devvit-debug',
    ElysiumInstallationOverride: 'devvit-elysium-installation-override',
    FeaturedGames: 'devvit-featured-games',
    GQLHost: 'devvit-gql-host',
    Installation: 'devvit-installation',
    ModPermissions: 'devvit-mod-permissions',
    Post: 'devvit-post',
    PostAuthor: 'devvit-post-author',
    PostData: 'devvit-post-data',
    RemoteHostname: 'devvit-remote-hostname',
    StreamID: 'devvit-stream-id',
    Subreddit: 'devvit-subreddit',
    SubredditName: 'devvit-subreddit-name',
    User: 'devvit-user',
    Username: 'devvit-user-name',
    UserSnoovatarUrl: 'devvit-user-snoovatar-url',
    UserAgent: 'devvit-user-agent',
    Version: 'devvit-version',
    Language: 'devvit-accept-language',
    Timezone: 'devvit-accept-timezone',
    Traceparent: 'traceparent',
});
/** Prefix common to all Devvit system headers. */
export const headerPrefix = 'devvit-';
/** See DevvitGlobal and ContextDebugInfo. */
export var AppDebug;
(function (AppDebug) {
    /** Enable debug logging for blocks. */
    AppDebug["Blocks"] = "blocks";
    /**
     * Log the entire reified blocks JSX/XML tree on each render. Eg:
     *
     *   <hstack><text>hi world</text></hstack>
     */
    AppDebug["EmitSnapshots"] = "emitSnapshots";
    /** Log app state changes. */
    AppDebug["EmitState"] = "emitState";
    /** Enable debug logging for realtime and useChannel() hook. */
    AppDebug["Realtime"] = "realtime";
    /** Enable runtime and dispatcher logging. */
    AppDebug["Runtime"] = "runtime";
    /** Enable debug logging for devvit-surface and dispatcher. */
    AppDebug["Surface"] = "surface";
    /** Enable debug logging for the useAsync() hook family. */
    AppDebug["UseAsync"] = "useAsync";
    /** Enable debug logging for payments APIs */
    AppDebug["Payments"] = "payments";
    /** Enable bootstrap logging */
    AppDebug["Bootstrap"] = "bootstrap";
    /** WebView debug logs */
    AppDebug["WebView"] = "webView";
})(AppDebug || (AppDebug = {}));
